plugins/intrinsics/src/main/java/org/qbicc/plugin/intrinsics/core/CoreIntrinsics.java

```diff
@@ -1210,9 +1210,18 @@ static void registerOrgQbiccRuntimeCNativeIntrinsics(final CompilationContext ct
 
         InstanceIntrinsic identity = (builder, instance, target, arguments) -> instance;
 
-        intrinsics.registerIntrinsic(ptrDesc, "deref", MethodDescriptor.synthesize(classContext, nObjDesc, List.of()), identity);
         intrinsics.registerIntrinsic(ptrDesc, "asArray", MethodDescriptor.synthesize(classContext, ArrayTypeDescriptor.of(classContext, nObjDesc), List.of()), identity);
 
+        InstanceIntrinsic deref = (builder, instance, target, arguments) -> {
+            MemoryAtomicityMode mode = MemoryAtomicityMode.UNORDERED;
+            if (instance.getType() instanceof PointerType pt && pt.getPointeeType() instanceof CompoundType) {
+                /* structures cannot be loaded atomically */
+                mode = MemoryAtomicityMode.NONE;
+            }
+            return builder.load(builder.pointerHandle(instance), mode);
+        };
+        intrinsics.registerIntrinsic(ptrDesc, "deref", MethodDescriptor.synthesize(classContext, nObjDesc, List.of()), deref);
```

Comment:
As long as we're in here and making at least one more change (below), please go ahead and rename this method from `deref` to `loadUnshared` and make it always use `MemoryAtomicityMode.NONE`.  Over this week's break (US Thanksgiving holiday) I plan to do a fairly extensive writeup of the JMM and how it applies to us and to LLVM and I expect we'll be adding more `load*` variants here afterwards.

Reply:
Sounds good. What is the meaning of Unshared for this?

Reply:
Also derefAssign -> loadUnsharedAssign? Not sure if there are plans to use derefAssign.

Reply:
`derefAssign` will become `storeUnshared`.  `Unshared` refers to the memory semantics of the `NONE` memory ordering mode.

I'm working on a much more detailed writeup, but the quick version is:

* `MemoryAtomicityMode.NONE` -> "Unshared"
* `MemoryAtomicityMode.UNORDERED` -> "Plain" (per JMM) or "Unordered" (per LLVM)
* `MemoryAtomicityMode.MONOTONIC` -> "Opaque" (per JMM) or "Relaxed" (per C/C++) or "Monotonic" (per LLVM)

The others don't really have other names, but I'm trying to recenter around the JMM terminology whenever possible.  We'll see at least the skeleton API for accessing many more modes via the pointer API soon.

Reply:
ooh right. store should have been obvious. Thanks for the details. I've updated the naming then and also rebased to resolve the merge conflict.